crates/handlers/src/upstream_oauth2/callback.rs

```diff
@@ -209,35 +213,52 @@ pub(crate) async fn get(
         redirect_uri,
     };
 
-    let id_token_verification_data = JwtVerificationData {
+    let verification_data = JwtVerificationData {
         issuer: &provider.issuer,
         jwks: &jwks,
         // TODO: make that configurable
         signing_algorithm: &mas_iana::jose::JsonWebSignatureAlg::Rs256,
         client_id: &provider.client_id,
     };
 
-    let (response, id_token) =
+    let (response, id_token_map) =
         mas_oidc_client::requests::authorization_code::access_token_with_authorization_code(
             &client,
             client_credentials,
             lazy_metadata.token_endpoint().await?,
             code,
             validation_data,
-            Some(id_token_verification_data),
+            Some(verification_data),
             clock.now(),
             &mut rng,
         )
         .await?;
 
-    let (_header, id_token) = id_token.ok_or(RouteError::MissingIDToken)?.into_parts();
+    let (_header, id_token) = id_token_map
+        .clone()
+        .ok_or(RouteError::MissingIDToken)?
+        .into_parts();
 
-    let env = {
-        let mut env = environment();
-        env.add_global("user", minijinja::Value::from_serialize(&id_token));
-        env
+    let fetch_userinfo_endpoint = match provider.user_profile_method {
+        UpstreamOAuthProviderUserProfileMethod::Auto => !provider.scope.contains("openid"),
+        UpstreamOAuthProviderUserProfileMethod::UserinfoEndpoint => true,
     };
 
+    let userinfo = if fetch_userinfo_endpoint {
+        mas_oidc_client::requests::userinfo::fetch_userinfo(
+            &client,
+            lazy_metadata.userinfo_endpoint().await?,
+            response.access_token.as_str(),
+            Some(verification_data),
+            &id_token_map.ok_or(RouteError::MissingIDToken)?,
+        )
+        .await?
+    } else {
+        HashMap::new()
+    };
+
+    let env = environment(id_token, userinfo.clone());
```

Comment:
I think I used the `environment` here the wrong way. In #3521, I made the environment 'static', and rather passed the dynamic context as template context. I'll let you know once that PR is merged so you can rebase your work on it

Reply:
#3521 has been merged, so instead what I suggest is adding the `userinfo_claims` in the template context here: https://github.com/element-hq/matrix-authentication-service/blob/7296364cd3ea9abe06ae4bb67bffeb7fc28a5978/crates/handlers/src/upstream_oauth2/template.rs#L24

and then make sure it can be used in the template here: https://github.com/element-hq/matrix-authentication-service/blob/7296364cd3ea9abe06ae4bb67bffeb7fc28a5978/crates/handlers/src/upstream_oauth2/template.rs#L54-L79